packages/smart/contracts/rewards/MasterChef.sol

```diff
@@ -0,0 +1,272 @@
+    }
+    // The REWARD TOKEN!
+    IERC20 private rewardsToken;
+    // Info of each pool.
+    PoolInfo[] public poolInfo;
+    // Info of each user that stakes LP tokens.
+    mapping(uint256 => mapping(address => UserInfo)) public userInfo;
+    // Total allocation points. Must be the sum of all allocation points in all pools.
+    uint256 public totalAllocPoint = 0;
+
+    mapping(address => bool) private approvedAMMFactories;
+
+    modifier onlyApprovedAMMFactories() {
+        require(approvedAMMFactories[msg.sender], "Caller must be approved.");
+        _;
+    }
+
+    event Deposit(address indexed user, uint256 indexed pid, uint256 amount);
+    event Withdraw(address indexed user, uint256 indexed pid, uint256 amount);
+    event EmergencyWithdraw(address indexed user, uint256 indexed pid, uint256 amount);
+
+    constructor(IERC20 _rewardsToken) {
+        rewardsToken = _rewardsToken;
+    }
+
+    function trustAMMFactory(address _ammFactory) public onlyOwner {
+        approvedAMMFactories[_ammFactory] = true;
+    }
+
+    function untrustAMMFactory(address _ammFactory) public onlyOwner {
+        delete approvedAMMFactories[_ammFactory];
+    }
+
+    function poolLength() external view returns (uint256) {
+        return poolInfo.length;
+    }
+
+    // Add a new lp to the pool. Can only be called by the owner.
+    // XXX DO NOT add the same LP token more than once. Rewards will be messed up if you do.
+    function add(
+        IERC20 _lpToken,
+        uint256 _rewardsPeriod,
+        uint256 _rewardsPerPeriod
+    ) public onlyApprovedAMMFactories returns (uint256 _nextPID) {
+        _nextPID = poolInfo.length;
+        poolInfo.push(
+            PoolInfo({
+                accRewardsPerShare: 0,
+                beginTimestamp: block.timestamp,
+                lpToken: _lpToken,
+                rewardsPeriod: _rewardsPeriod,
+                rewardsPerPeriod: _rewardsPerPeriod,
+                lastRewardTimestamp: block.timestamp
+            })
+        );
+    }
+
+    // Return percentage of period that has elapsed in terms of BONE.
+    function getPercentageOfRewardsForPeriod(uint256 _pid) public view returns (uint256) {
+        PoolInfo storage _pool = poolInfo[_pid];
+        uint256 _fromTimestamp = block.timestamp;
+
+        if (
+            // Rewards have not started yet.
+            _pool.beginTimestamp > _fromTimestamp ||
+            // Not sure how this happens but it is accounted for in the original master chef contract.
+            _pool.lastRewardTimestamp > _fromTimestamp
+        ) {
+            return 0;
+        }
+
+        uint256 _rewardsPeriodInSeconds = _pool.rewardsPeriod * 1 days;
+        uint256 _rewardPeriodEndTimestamp = _rewardsPeriodInSeconds + _pool.beginTimestamp + 1;
+
+        // Rewards are over for this pool. No more rewards have accrued.
+        if (_pool.lastRewardTimestamp >= _rewardPeriodEndTimestamp) {
```

Comment:
_pool.lastRewardTimestamp is the timestamp of the block  when the rewards were updated using updatePool, i.e., from a deposit or widthdraw.  block times on Matic are small 2s  so is this comparison accurate enough. Also this inequality appears to need updatePool to be called ( withdrawInternal or depositInternal) first to set _pool.lastRewardTimestamp but I don't see a guarantee this will happen

   I

Reply:
I see your point. Cannot call a method (`updatePool`) which mutates state within a view method. The caller will be responsible for updating this if they are interested in an up to date calculation.